### PR TITLE
Re-land "Ensure flutter build apk --release optimizes+shrinks platform code"

### LINKED ISSUE #136879: Flutter release builds (via `flutter build apk --release`) aren't optimized
It seems flutter release builds result in APKs that have dex in them (from java/kotlin) that wasn't optimized.

This can be seen as follows:
```
% flutter create hello && cd hello

### Build the APK
hello %  flutter build apk --release

### Examine the R8/Proguard rules used for tree shaking
hello % cat build/app/outputs/mapping/release/configuration.txt
...
# Optimization is turned off by default. ...
-dontoptimize
...


### Examine size of resulting DEX bytecode
hello % unzip -l build/app/outputs/flutter-apk/app-release.apk | grep classes.dex
   689484  1981-01-01 01:01   classes.dex
```

The root cause is that flutter's gradle file explicitly tells the R8 to use unoptimized android proguard rules.

If we fix this via:
```diff
diff --git a/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy b/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
index f3c1a4145c..f5950024e2 100644
--- a/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
+++ b/packages/flutter_tools/gradle/src/main/groovy/flutter.groovy
@@ -308,7 +308,7 @@ class FlutterPlugin implements Plugin<Project> {
                     shrinkResources isBuiltAsApp(project)
                     // Fallback to `android/app/proguard-rules.pro`.
                     // This way, custom Proguard rules can be configured as needed.
-                    proguardFiles project.android.getDefaultProguardFile("proguard-android.txt"), flutterProguardRules, "proguard-rules.pro"
+                    proguardFiles project.android.getDefaultProguardFile("proguard-android-optimize.txt"), flutterProguardRules, "proguard-rules.pro"
                 }
             }
         }
```

and rebuild we get the following:
```
hello % flutter build apk --release

hello % unzip -l build/app/outputs/flutter-apk/app-release.apk | grep classes.dex
   512376  1981-01-01 01:01   classes.dex
```

=> The `classes.dex` file shrinks by **25%** (`689484` -> `512376`).
=> Bigger apps that use various java/kotlin plugins may benefit quite significantly (compared with the hello world)

/cc @reidbaker

### PR DESCRIPTION
**Pull request by gmackall**
Re-lands https://github.com/flutter/flutter/pull/136880, fixes https://github.com/flutter/flutter/issues/136879.

Additions to/things that are different from the original PR:
- Adds an entry to `gradle_errors.dart` that tells people when they run into the R8 bug because of using AGP 7.3.0 (https://issuetracker.google.com/issues/242308990).
- Previous PR moved templates off of AGP 7.3.0.
- Packages repo has been moved off AGP 7.3.0 (https://github.com/flutter/packages/pull/7432).

Also, unrelatedly:
- Deletes an entry in `gradle_errors.dart` that informed people to build with `--no-shrink`. This flag [doesn't do anything](https://github.com/flutter/website/pull/11022#issuecomment-2297294421), so it can't be the solution to any error.
- Uniquely lowers the priority of the `incompatibleKotlinVersionHandler`. This is necessary because the ordering of the errors doesn't fully determine the priority of which handler we decide to use, but also the order of the log lines. The kotlin error lines often print before the other error lines, so putting it last in the list of handlers isn't sufficient to lower it to be the lowest priority handler.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter_tools/lib/src/android/gradle_errors.dart [hunk 1]
```diff
@@ -642,6 +620,14 @@ final GradleHandledError couldNotOpenCacheDirectoryHandler = GradleHandledError(
   eventLabel: 'could-not-open-cache-directory',
 );
 
+
+String _getAgpLocation(FlutterProject project) {
+  return ' The version of AGP that your project uses is likely'
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/android/gradle_errors.dart:625]: nit: You could consider multi-line strings:
```dart
return '''
The version of AGP that your project uses is likely
${project.android.settingsGradleFile.path}
in the 'plugins' closure.
Alternatively, ...
''';
```

(also other places)

File: packages/flutter_tools/lib/src/android/gradle_errors.dart [hunk 2]
```diff
@@ -664,3 +647,23 @@ final GradleHandledError incompatibleCompileSdk35AndAgpVersionHandler = GradleHa
   },
   eventLabel: 'incompatible-compile-sdk-and-agp',
 );
+
+@visibleForTesting
+final GradleHandledError r8DexingBugInAgp73Handler = GradleHandledError(
+  test: (String line) => line.contains('com.android.tools.r8.internal.Y10: Unused argument with users'),
```
Comment (mkustermann) [inline packages/flutter_tools/lib/src/android/gradle_errors.dart:653]: You may not want to match `Y10` as that identifier could be different between minor versions of R8 (it's very likely a obfuscated class name)

**Review by mkustermann — APPROVED** [review]
LGTM, thanks for working on this, @gmackall !

File: packages/flutter_tools/lib/src/android/gradle_errors.dart (hunk 1 above)
Comment (gmackall) [reply packages/flutter_tools/lib/src/android/gradle_errors.dart]: Changed to multi line strings

**Comment by gmackall** [comment]
Updated for review, and to make sure that `incompatibleKotlinVersionHandler` never overrides another handler.

Summary of why this is necessary is that we test each log line as it comes in, and the lines that trigger `incompatibleKotlinVersionHandler` come before most others. So putting it last in the list of handlers doesn't work. In fact, the order of the list will almost never matter. 

Priority of handlers is determined by order of log lines, with only ties being broken by the order of the list of handlers.

**Review by mkustermann — APPROVED** [review]
Still LGTM

**Review by reidbaker — COMMENTED** [review]
Consider adding some of the contents of https://github.com/flutter/flutter/pull/153868#issuecomment-2305596169 into the description of gradleErrors.

**Comment by zanderso** [comment]
FYI: This change regressed release build times on Windows hosts by ~30 seconds: https://flutter-flutter-perf.skia.org/e/?begin=1722984359&end=1725384267&keys=X3ba6023ddb62c0aa3cb1d6d43943ab9a&requestType=0&selected=commit%3D42295%26name%3D%252Carch%253Dintel%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253Dmokey%252Cdevice_version%253Dnone%252Chost_type%253Dwin%252Csub_result%253Drelease_full_compile_millis%252Ctest%253Dflutter_gallery_win__compile%252C&xbaroffset=42295.

The regression on more modern M1 mac minis was less severe in absolute terms. No regression was detected on Linux.

Just flagging this to ensure folks are aware of this trade-off.

**Comment by zanderso** [comment]
Improvements in binary size: https://flutter-flutter-perf.skia.org/e/?begin=1724422710&end=1724435825&keys=Xd5e5f4ba606468149799fd7e089344e5&num_commits=50&request_type=1&xbaroffset=42295

**Comment by reidbaker** [comment]
We expect an increase in build time when it comes to optimizing code size. The benefits and costs grow as the size/complexity of your code grows.

**Pull request merged**